### ISSUE TITLE
Why hmean is alway 0.000 when train toy dataset by textsnake or dbnet

### ISSUE THREAD

**Issue by xiefeifeihu** [open-mmlab/mmocr#1109]
My config is like this:

_base_ = [
    '../configs/_base_/default_runtime.py',
    '../configs/_base_/schedules/schedule_sgd_1200e.py',
    '../configs/_base_/det_models/dbnet_r18_fpnc.py',
    '../configs/_base_/det_datasets/toy_data.py',
    '../configs/_base_/det_pipelines/dbnet_pipeline.py'
]

root = 'tests/data/toy_dataset'
train = dict(
    type='TextDetDataset',
    img_prefix=f'{root}/imgs',
    ann_file=f'{root}/instances_test.txt',
    loader=dict(
        type='AnnFileLoader',
        repeat=4,
        file_format='txt',
        parser=dict(
            type='LineJsonParser',
            keys=['file_name', 'height', 'width', 'annotations'])),
    pipeline=None,
    test_mode=False)
train_list = [train]
test_list = {{_base_.test_list}}

train_pipeline_r18 = {{_base_.train_pipeline_r18}}
test_pipeline_1333_736 = {{_base_.test_pipeline_1333_736}}

data = dict(
    samples_per_gpu=4,
    workers_per_gpu=4,
    val_dataloader=dict(samples_per_gpu=1),
    test_dataloader=dict(samples_per_gpu=1),
    train=dict(
        type='UniformConcatDataset',
        datasets=train_list,
        pipeline=train_pipeline_r18),
    val=dict(
        type='UniformConcatDataset',
        datasets=test_list,
        pipeline=test_pipeline_1333_736),
    test=dict(
        type='UniformConcatDataset',
        datasets=test_list,
        pipeline=test_pipeline_1333_736))

# evaluation = dict(interval=100, metric='hmean-iou')
evaluation = dict(interval=1, metric=["hmean-iou"], save_best="0_hmean-iou:hmean", rule="greater")
checkpoint_config = dict(interval=1, max_keep_ckpts=1)
runner = dict(type='EpochBasedRunner', max_epochs=10)

log_config = dict(
    interval=1,
    hooks=[
        dict(type='TextLoggerHook'),
        dict(type='TensorboardLoggerHook')])
work_dir = 'work_dirs/textdet/dbnet/toy_dataset'

And the log is:
[20220624_135006.log](https://github.com/open-mmlab/mmocr/files/8973570/20220624_135006.log)

**Comment by xiefeifeihu** [comment]
I debug the code in Postprocessor, I found the boundary_result is empty list. Both TextSnakePostprocessor and DBPostprocessor.

Are there any mistakes in my config file?

**Comment by Mountchicken** [comment]
Hi @xiefeifeihu 
Training only on the toy dataset may not be enough for the network to convergence.

**Comment by xiefeifeihu** [comment]
> Hi @xiefeifeihu Training only on the toy dataset may not be enough for the network to convergence.

怎么着都不至于是0吧，训练500个epoch都是0.00。症结应该不是数据量，因为我用自己的数据集训练mmocr也是这样，但训练mmdet和mmseg却是正常的。

**Comment by gaotongxiao** [comment]
Text detection models are empirically very sensitive to learning rate and batch size. I noticed the loss is 5x greater than that of converged ones.

Also you might need to tune postprocessors to fit different scenarios (e.g. https://github.com/open-mmlab/mmocr/issues/376).

**Comment by gaotongxiao** [comment]
And I just modified `configs/textdet/dbnet/dbnet_r18_fpnc_1200e_icdar2015.py` to use the following dataset config, while setting `samples_per_gpu=workers_per_gpu=4`.

```
root = 'tests/data/toy_dataset'

# dataset with type='TextDetDataset'
train1 = dict(
    type='TextDetDataset',
    img_prefix=f'{root}/imgs',
    ann_file=f'{root}/instances_test.txt',
    loader=dict(
        type='AnnFileLoader',
        repeat=4,
        file_format='txt',
        parser=dict(
            type='LineJsonParser',
            keys=['file_name', 'height', 'width', 'annotations'])),
    pipeline=None,
    test_mode=False)

test = dict(
    type='TextDetDataset',
    img_prefix=f'{root}/imgs',
    ann_file=f'{root}/instances_test.txt',
    loader=dict(
        type='AnnFileLoader',
        repeat=1,
        file_format='txt',
        parser=dict(
            type='LineJsonParser',
            keys=['file_name', 'height', 'width', 'annotations'])),
    pipeline=None,
    test_mode=True)

train_list = [train1]
test_list = [test]
```

The hmean is 0.6857 at 100th epoch. The default configuration works fine on my side.

**Comment by xiefeifeihu** [comment]
> Text detection models are empirically very sensitive to learning rate and batch size. I noticed the final loss in your log is 5x greater than that of converged ones.
> 
> Also you might need to tune postprocessors to fit different scenarios (e.g. #376).

Thank you very much. Could you upload your full configuration, log, and dependencies? I want to compare differences to troubleshoot my problems.

**Comment by gaotongxiao** [comment]
@xiefeifeihu Sure. Check this out: https://gist.github.com/gaotongxiao/4e326a63cb1274c6b107e72f53167b37

Note that I did not do any other modifications to default configs except for the aforementioned ones.

**Comment by xiefeifeihu** [comment]
@gaotongxiao 
Hi,I make my configuration same with you，and create a new environment same with you（Python: 3.8.13、PyTorch: 1.11.0、TorchVision: 0.12.0+cu102、OpenCV: 4.5.4、MMCV: 1.4.8）。Then train the toy dataset in both old and new environment。

I found that the hmean metric is randomness independent of the environment。Both environment have experienced hmean>0 after 200th epoch, and hmean==0.000 after 600th epoch.Please see the log:

hmean>0 after 200th epoch——
`
2022-06-29 10:39:30,008 - mmocr - INFO - Epoch [200][10/10]	lr: 5.946e-03, eta: 1:19:22, time: 0.203, data_time: 0.036, memory: 1793, loss_prob: 2.2312, loss_db: 0.3437, loss_thr: 0.8614, loss: 3.4363
2022-06-29 10:39:30,117 - mmocr - INFO - Saving checkpoint at 200 epochs
2022-06-29 10:39:30,952 - mmocr - INFO - 
Evaluating tests/data/toy_dataset/instances_test.txt with 10 images now
2022-06-29 10:39:30,953 - mmocr - INFO - Evaluating hmean-iou...
2022-06-29 10:39:30,986 - mmocr - INFO - thr 0.30, recall: 0.381, precision: 0.320, hmean: 0.348
2022-06-29 10:39:31,005 - mmocr - INFO - thr 0.40, recall: 0.381, precision: 0.533, hmean: 0.444
2022-06-29 10:39:31,007 - mmocr - INFO - thr 0.50, recall: 0.000, precision: 0.000, hmean: 0.000
2022-06-29 10:39:31,009 - mmocr - INFO - thr 0.60, recall: 0.000, precision: 0.000, hmean: 0.000
2022-06-29 10:39:31,011 - mmocr - INFO - thr 0.70, recall: 0.000, precision: 0.000, hmean: 0.000
2022-06-29 10:39:31,013 - mmocr - INFO - thr 0.80, recall: 0.000, precision: 0.000, hmean: 0.000
2022-06-29 10:39:31,015 - mmocr - INFO - thr 0.90, recall: 0.000, precision: 0.000, hmean: 0.000
2022-06-29 10:39:31,015 - mmocr - INFO - Exp name: dbnet_r18_fpnc_1200e_toy_dataset.py
2022-06-29 10:39:31,016 - mmocr - INFO - Epoch(val) [200][10]	0_hmean-iou:recall: 0.3810, 0_hmean-iou:precision: 0.5333, 0_hmean-iou:hmean: 0.4444
`

hmean==0.000 after 600th epoch——
`
2022-06-29 14:31:53,089 - mmocr - INFO - Epoch [600][10/10]	lr: 3.757e-03, eta: 0:46:50, time: 0.186, data_time: 0.031, memory: 1792, loss_prob: 2.8343, loss_db: 0.1428, loss_thr: 0.4631, loss: 3.4403
2022-06-29 14:31:53,170 - mmocr - INFO - Saving checkpoint at 600 epochs
2022-06-29 14:31:53,901 - mmocr - INFO - 
Evaluating tests/data/toy_dataset/instances_test.txt with 10 images now
2022-06-29 14:31:53,901 - mmocr - INFO - Evaluating hmean-iou...
2022-06-29 14:31:53,903 - mmocr - INFO - thr 0.30, recall: 0.000, precision: 0.000, hmean: 0.000
2022-06-29 14:31:53,905 - mmocr - INFO - thr 0.40, recall: 0.000, precision: 0.000, hmean: 0.000
2022-06-29 14:31:53,907 - mmocr - INFO - thr 0.50, recall: 0.000, precision: 0.000, hmean: 0.000
2022-06-29 14:31:53,909 - mmocr - INFO - thr 0.60, recall: 0.000, precision: 0.000, hmean: 0.000
2022-06-29 14:31:53,911 - mmocr - INFO - thr 0.70, recall: 0.000, precision: 0.000, hmean: 0.000
2022-06-29 14:31:53,912 - mmocr - INFO - thr 0.80, recall: 0.000, precision: 0.000, hmean: 0.000
2022-06-29 14:31:53,914 - mmocr - INFO - thr 0.90, recall: 0.000, precision: 0.000, hmean: 0.000
2022-06-29 14:31:53,914 - mmocr - INFO - Exp name: dbnet_r18_fpnc_1200e_toy_dataset.py
`

**Comment by xiefeifeihu** [comment]
@gaotongxiao 
I trained my own dataset many times，hmean is always 0.000。but the model generated can do a little inference，A few ocr results are still correct。such as “{'filename': 'SF_F_014', 'text': ['zoco-12-is', '方式式', '付0210118470', '170.', 'SF1038052605558']}”

请问一下，标注是否也有些技巧。比如图片上文字较多，但感兴趣的很少（可能仅占5%），那么其他不感兴趣的内容是否需要标注？若需要标注是否标注为###即可？

**Comment by xiefeifeihu** [comment]
@balandongiv no. hmean is 0.348 after 200th epoch when i trained my own dataset.

**Comment by balandongiv** [comment]
Did you modify the `epsilon`? @xiefeifeihu

**Comment by xiefeifeihu** [comment]
@balandongiv no

**Comment by gaotongxiao** [comment]
Sorry for the late reply.

And for your question, you would have to tune the hyperparameters according to the datasets, but we are not able to provide practical guidance since situations usually vary case by case. You can visualize the model's results to gain a better understanding of the issue.

As per annotating data, you're right, marking "do-not-care" regions with "###" tells MMOCR to skip those regions during training, and therefore the model can be more "focused" on the targrts.

**Comment by xiefeifeihu** [comment]
@gaotongxiao OK, Thank you. I close this issue.

Hope the community can provide some training of MMOCR.
